File: Unit D/Dprogramming/Java/Dprogramming/src/TilingDino.java
/**
 * CS4102 Spring 2022 -- Unit D Programming
 *********************************
 * Collaboration Policy: You are encouraged to collaborate with up to 3 other
 * students, but all work submitted must be your own independently written
 * solution. List the computing ids of all of your collaborators in the comment
 * at the top of your java or python file. Do not seek published or online
 * solutions for any assignments. If you use any published or online resources
 * (which may not include solutions) when completing this assignment, be sure to
 * cite them. Do not submit a solution that you are unable to explain orally to a
 * member of the course staff.
 *********************************
 * Your Computing ID: wyz5rge
 * Collaborators: N/A
 * Sources: Introduction to Algorithms, Cormen
 * 			Discrete Mathematics, Reza Zadeh, Standord University
 * 			
 **************************************/

import java.util.ArrayList;
import java.util.List;

import org.jgrapht.*;
import org.jgrapht.alg.flow.EdmondsKarpMFImpl;
import org.jgrapht.alg.interfaces.FlowAlgorithm;
import org.jgrapht.alg.interfaces.FlowAlgorithm.FlowImpl;
import org.jgrapht.alg.interfaces.MatchingAlgorithm;
import org.jgrapht.alg.interfaces.MatchingAlgorithm.MatchingImpl;
import org.jgrapht.alg.interfaces.MaximumFlowAlgorithm;
import org.jgrapht.alg.matching.MaximumWeightBipartiteMatching;
import org.jgrapht.graph.*;
import org.jgrapht.nio.*;
import org.jgrapht.nio.dot.*;
import org.jgrapht.traverse.*;

import java.io.*;
import java.net.*;
import java.util.*;

public class TilingDino {

    /**
     * This is the method that should set off the computation
     * of tiling dino.  It takes as input a list lines of input
     * as strings.  You should parse that input, find a tiling,
     * and return a list of strings representing the tiling
     *
     * @return the list of strings representing the tiling
     */
    public List<String> compute(List<String> fileLines) 
    {
    	List<String> result = new ArrayList<>();
        
    	/*
    	 * I am going to take a chekerboard approach, assigning pixels we want to tile to be
    	 * either white or black, then finding a bipartite matching based off the flow newtwork
    	 * created by that
    	 */
    	
    	int pixels = 0;
    	int rows = fileLines.size();
    	int columns = fileLines.get(0).length();
    	
    	// DefaultDirectedGraph instead of DefaultDirectedWeightGraph because default edges are treated as having a weight of 1
    	Graph<String, DefaultWeightedEdge> g = new SimpleDirectedWeightedGraph<>(DefaultWeightedEdge.class);
    	
    	// initialize source and sink nodes
    	g.addVertex("source");
    	g.addVertex("sink");
    	
    	// nested for-loop that populates the graph by traversing the input "grid"
    	for(int r = 0; r < rows; r++)
    	{
    		for(int c = 0; c < columns; c++)
    		{
    			if(fileLines.get(r).charAt(c) == '#')
    			{	
    				pixels++;
    				
    				/*
    				if(r % 2 == 0)
    				{
    					if(c % 2 == 0)
    					{
    						g.addVertex("w " + Integer.toString(c) + " " + Integer.toString(r));
    						// add edge between source node and node representing "white" pixel
    						g.addEdge("source", "w " + Integer.toString(c) + " " + Integer.toString(r));
    					}
    					else
    					{
    						g.addVertex("b " + Integer.toString(c) + " " + Integer.toString(r));
    						// add edge between "black" pixel node and sink node
    						g.addEdge("b " + Integer.toString(c) + " " + Integer.toString(r), "sink");
    					}
    				}
    				else
    				{
    					if(c % 2 == 0)
    					{
    						g.addVertex("b " + Integer.toString(c) + " " + Integer.toString(r));
    						g.addEdge("b " + Integer.toString(c) + " " + Integer.toString(r), "sink");
    					}
    					else
    					{
    						g.addVertex("w " + Integer.toString(c) + " " + Integer.toString(r));
    						g.addEdge("source", "w " + Integer.toString(c) + " " + Integer.toString(r));
    					}
    				}
    				*/
    				
    				// the following code block is the same as above, just in fewer lines
    				if(r % 2 == c % 2)
    				{
    					// instatiate the vertices with c first to stay consistent with the assignment description
    					g.addVertex("w " + Integer.toString(c) + " " + Integer.toString(r));
    					// add edge between source node and node representing "white" pixel
						g.addEdge("source", "w " + Integer.toString(c) + " " + Integer.toString(r));
    				}
    				else
    				{
    					g.addVertex("b " + Integer.toString(c) + " " + Integer.toString(r));
    					// add edge between "black" pixel node and sink node
						g.addEdge("b " + Integer.toString(c) + " " + Integer.toString(r), "sink");
    				}
    			}
    		}
    	}

    	/*
    	 * return impossible if there aren't an even number of pixels - obviously no perfect bartite matching
    	 * can tell this works as intended - test1 returns impossible and test2 does not
    	 */
    	if(pixels % 2 != 0)
    	{
    		result.add("impossible");
    		return result;
    	}
    	
    	/*
    	 * traverse input agian to instatiate edges between black and white edges
    	 * i tried to do all of this within the same nested for-loop but it was hard to do that 
    	 * when some nodes weren't created yet
    	 */
    	for(int r = 0; r < rows; r++)
    	{
    		for(int c = 0; c < columns; c++)
    		{
    			if(fileLines.get(r).charAt(c) == '#')
    			{
    				/*
    				 *  since we're iterating left-right, up-down, we only need to care about
    				 *  the pixels below and to the right of the current pixel
    				 */
    				if(c + 1 < columns && fileLines.get(r).charAt(c + 1) == '#')
    				{
    					if(r % 2 == c % 2)
    						g.addEdge("w " + Integer.toString(c) + " " + Integer.toString(r) , "b " + Integer.toString(c + 1) + " " + Integer.toString(r));
    					else
    						g.addEdge("w " + Integer.toString(c + 1) + " " + Integer.toString(r) , "b " + Integer.toString(c) + " " + Integer.toString(r));
    				}
    				if(r + 1 < rows && fileLines.get(r + 1).charAt(c) == '#')
    				{
    					if(r % 2 == c % 2)
    						g.addEdge("w " + Integer.toString(c) + " " + Integer.toString(r) , "b " + Integer.toString(c) + " " + Integer.toString(r + 1));
    					else
    						g.addEdge("w " + Integer.toString(c) + " " + Integer.toString(r + 1) , "b " + Integer.toString(c) + " " + Integer.toString(r));
    				}
    			}
    		}
    	}
    	
    	/*
    	 * the algorithm only reaches this point if the number of pixels is even,
    	 * thus the number of pairs must be the number of pixels divided by two
    	 */
    	int pairs = pixels / 2;
    	
    	
    	/*
    	 * create set of all edges in the bipartite graph,
    	 * declare matching algorithm class, set a new Set structure to the getEdges method
    	 */
    	/*
    	Set<DefaultEdge> edges = g.edgeSet();
    	
    	
    	
		MatchingAlgorithm.MatchingImpl<String, DefaultEdge> match = new MatchingImpl<String, DefaultEdge>(g, edges, pairs);
		Set<DefaultEdge> matching = match.getEdges();
		*/
		/*
		for(DefaultEdge e : matching) 
		{
			System.out.println(e.toString());
		}
		*/
		
		/*
		 * print loop to see if we got the right edges
		 * i think it works??
		 * okay yeah this definintely works, it output the source for all of the expected nodes
		 * now time to figure out how to convert all of these to strings...		
		for(DefaultEdge e : matching) 
		{
			System.out.println("source: " + g.getEdgeSource(e) + "; target: " + g.getEdgeTarget(e));
		}
		*/
		
		/*
		// we have the correct edges, but it also includes edges connecting to source and sink nodes, we need to remove those
		ArrayList<String> list = new ArrayList<String>();
		
		for(DefaultEdge e : matching)
		{
			String s = e.toString();
			if(!s.contains("source") && !s.contains("sink"))
			{
				list.add(s);
			}
		}
		
		// System.out.println(list);
		
		ArrayList<String> list2 = new ArrayList<String>();
		for(int i = 0; i < list.size(); i++)
		{
			list2.add(list.get(i).replaceAll("\\D", " "));
		}
		
		// System.out.println(list2);
		
		ArrayList<String> list3 = new ArrayList<String>();
		for(int i = 0; i < list2.size(); i++)
		{
			list3.add(list2.get(i).replaceAll("\\s+", " "));
		}
				
		for(int i = 0; i < list3.size(); i++)
		{
			System.out.println(list3.get(i));;
		}
		*/

    	EdmondsKarpMFImpl<String, DefaultWeightedEdge> maxFlow = new EdmondsKarpMFImpl<String, DefaultWeightedEdge>(g);
    	System.out.println(maxFlow.getMaximumFlow("source", "sink"));	
		
        return result;
    }   
}

File: Unit D/Dprogramming/Python/TilingDino.py
# CS4102 Spring 2022 -- Unit D Programming
#################################
# Collaboration Policy: You are encouraged to collaborate with up to 3 other
# students, but all work submitted must be your own independently written
# solution. List the computing ids of all of your collaborators in the comment
# at the top of your java or python file. Do not seek published or online
# solutions for any assignments. If you use any published or online resources
# (which may not include solutions) when completing this assignment, be sure to
# cite them. Do not submit a solution that you are unable to explain orally to a
# member of the course staff.
#################################
# Your Computing ID: wyz5rge
# Collaborators: N/A
# Sources: Introduction to Algorithms, Cormen
#################################

# originally started this assignment in Java, switched to Python because jgrapht is annoying to work with
# final submission, cleaned up some code

# having this package would be pretty useful...
import networkx

class TilingDino:
    def __init__(self):
        return

    # This is the method that should set off the computation
    # of tiling dino.  It takes as input a list lines of input
    # as strings.  You should parse that input, find a tiling,
    # and return a list of strings representing the tiling
    #
    # @return the list of strings representing the tiling
    def compute(self, lines):

        # just like in my java project, we're going to assigned tiled pixels in a checkerboard pattern,
        # then find a perfect bitartite matching based after generating the graph

        # this will be the array that the domino tilings will be added to
        solution = []

        # dimensions of picture
        rows = len(lines)
        columns = len(lines[0])

        g = networkx.DiGraph()

        # bipartite matching algorithm is solved by reducing to vertex-disjoint flow network
        g.add_node("source")
        g.add_node("sink")

        # count number of pixels that need to be tiled, and assign b/w based on position
        pixels = 0
        for r in range(rows):
            for c in range(columns):
                if lines[r][c] == "#":
                    pixels = pixels + 1

                    # very top left corner pixel will be "white"
                    # original implementation executed based on row and column modulus depending on even/oddness of row
                    # checking for equality serves the same purpose, but is implemented in half the number of lines
                    if r % 2 == c % 2:
                        # column number comes before row number to maintain consistency with the assignment description
                        pixelString = "w " + str(c) + " " + str(r)
                        g.add_node(pixelString)
                        g.add_edge("source", pixelString, capacity = 1)
                    else:
                        pixelString = "b " + str(c) + " " + str(r)
                        g.add_node(pixelString)
                        g.add_edge(pixelString, "sink", capacity = 1)

        # correctly counting number of pixels
        # print("pixels: " + pixels)

        # correctly returning impossible for odd nubmer of pixels
        if pixels % 2 == 1:
            solution.append("impossible")
            return solution

        # create edges that connect adjacent pixels
        # i tried to implement this in the nested for-loop above, but it's a little hard to create edges when certain vertices haven't been created yet
        for r in range(rows):
            for c in range(columns):
                if lines[r][c] == "#":
                    # check pixel below if curent pixel is not on bottom row
                    if c + 1 < columns and lines[r][c + 1] == "#":
                        if r % 2 == c % 2:
                            edgeSourceString = "w " + str(c) + " " + str(r)
                            edgeTargetString = "b " + str(c + 1) + " " + str(r)
                            g.add_edge(edgeSourceString, edgeTargetString, capacity = 1)
                            # print("edge added down white")
                        else:
                            edgeSourceString = "w " + str(c + 1) + " " + str(r)
                            edgeTargetString = "b " + str(c) + " " + str(r)
                            g.add_edge(edgeSourceString, edgeTargetString, capacity = 1)
                            # print("edge added down black")

                    # check pixel to the right if pixel is not on right border
                    if r + 1 < rows and lines[r + 1][c] == "#":
                        if r % 2 == c % 2:
                            edgeSourceString = "w " + str(c) + " " + str(r)
                            edgeTargetString = "b " + str(c) + " " + str(r + 1)
                            g.add_edge(edgeSourceString, edgeTargetString, capacity = 1)
                            # print("edge added right white")
                        else:
                            edgeSourceString = "w " + str(c) + " " + str(r + 1)
                            edgeTargetString = "b " + str(c) + " " + str(r)
                            g.add_edge(edgeSourceString, edgeTargetString, capacity = 1)
                            # print("edge added right black")

        maxFlow = networkx.maximum_flow(g, "source", "sink")
        maxFlow = list(maxFlow)

        # print("possible pairs: " + str(pixels / 2))
        # print("max flow: " + str(maxFlow[0]))

        # if networkx.maximum_flow_value(g, "source", "sink") == pixels / 2:
        if maxFlow[0] == pixels / 2:
            # print("printing solution")
            for source in maxFlow[1]:
                if source[0] == "w":
                    for target in maxFlow[1][source]:
                        if maxFlow[1][source][target] == 1:
                            sourceNode, c1, r1 = source.split(" ")
                            targetNote, c2, r2 = target.split(" ")
                            solution.append(c1 + " " + r1 + " " + c2 + " " + r2)
            return solution

        # originally just had this line return solution, but i realized that there's achance that a picture with an even number of pixels is still impossible to tile
        solution.append("impossible")
        return solution